salt/orch/force_removal.sls

```diff
@@ -0,0 +1,72 @@
+# This is a best effort orchestration. Any operation run on the minions by this orchestration
+# could fail, as we don't know the current state of the cluster at this point. The idea is to
+# try to clean as much as possible where possible, but in the end the only thing we can really
+# ensure is that the salt-key of the target is rejected.
+
+# must provide the node (id) to be removed in the 'target' pillar
+{%- set target = salt['pillar.get']('target') %}
+{%- set target_nodename = salt.saltutil.runner('mine.get', tgt=target, fun='nodename')[target] %}
+# services that will be called when trying to unregister as clean as possible
+{%- set services_to_stop = { "kubelet": ["stop", "remove-pre-reboot"], "etcd": ["remove-pre-stop-services", "remove-pre-reboot"] } %}
+# nodes that are responding
+{%- set up_nodes = salt.saltutil.runner('manage.up', tgt='P@roles:kube-* and not ' + target, expr_form='compound') %}
+
+# Try to unregister the machine from cluster services. It's important to keep every sls in
+# different toplevel stages, so each one will be tried even if previous fails for whatever
+# reason.
+{%- for service, sls_list in services_to_stop.items() %}
+  {%- for sls in sls_list %}
+stop-{{ service }}-{{ sls }}-stage:
+  salt.state:
+    - tgt: {{ target }}
+    - fail_minions: {{ target }}
+    - sls:
+        - {{ service }}.{{ sls }}
+  {%- endfor %}
+{%- endfor %}
+
+# Try to shutdown the target
+shutdown-target:
+  salt.function:
+    - tgt: {{ target }}
+    - name: cmd.run
+    - fail_minions: {{ target }}
+    - arg:
+      - systemctl poweroff
+    - kwarg:
+        bg: True
+
+# Even if trying to unregister the machine failed (e.g. `kubectl delete node` failed because
+# kubelet/stop.sls couldn't be executed on the target), we try to remove the node again from
+# the kubernetes cluster. This time from other working machine, and after we have done our
+# best to shut down the target. Do this from all the machines we can, best effort again.
+# Once that we managed to remove the node from the cluster, we stop.
+{%- set previous_node = none %}
+{%- for node in up_nodes %}
+try-to-unregister-{{ target }}-from-{{ node }}:
+  salt.function:
+    - tgt: {{ node }}
+    - name: cmd.run
+    - fail_minions: {{ node }}
+    - arg:
+        - kubectl --kubeconfig={{ pillar['paths']['kubeconfig'] }} delete node {{ target_nodename }}
+  {%- if previous_node %}
+    - onfail:
+        - try-to-unregister-{{ target }}-from-{{ previous_node }}
+  {%- endif %}
+  {%- set previous_node = node %}
+{%- endfor %}
+
```

Comment:
I think we should also remove the node from the etcd cluster...

Reply:
We are, we are running: https://github.com/kubic-project/salt/blob/master/salt/etcd/remove-pre-stop-services.sls

Reply:
That can fail, so you should do it in some other machine. Any machine where `etcdctl` can be run should do the job...

Reply:
Will fix